### PR TITLE
release sciencebeam version

### LINKED ISSUE #82: Release versions for ScienceBeam docker images
It seems to make sense to use release versions now that we encourage xPub to deploy their own ScienceBeam containers.

The versions could be managed like for releases of say `sciencebeam-utils`. If the version field from the package does not correspond to a git tag, perform the release. In our case push a docker image with the version.

Based on the discussions around the _libero_ project we could scope commit versioned containers with a suffix like `_unstable`. We could a bit later prune old commit versions from the current container image name.

In the future the code could be release on pypi as well but it doesn't seem useful at present.

/cc @giorgiosironi

### PR DESCRIPTION
**Pull request by de-code**
resolves #82

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -26,7 +34,15 @@ elifeLibrary {
             elifeGitMoveToBranch commit, 'master'
         }
 
-        stage 'Push image', {
+        stage 'Push release image', {
```
Comment (de-code) [inline Jenkinsfile]: seems to make sense as a separate step(?)

File: Jenkinsfile [hunk 2]
```diff
@@ -26,7 +34,15 @@ elifeLibrary {
...
+            isNew = sh(script: "git tag | grep v${candidateVersion}", returnStatus: true) != 0
+            if (isNew) {
+                def image = DockerImage.elifesciences(this, 'sciencebeam', commit)
+                image.tag(candidateVersion).push()
```
Comment (de-code) [inline Jenkinsfile]: should also push the latest release here

File: Jenkinsfile [hunk 3]
```diff
@@ -26,7 +34,15 @@ elifeLibrary {
...
+            }
+        }
+
+        stage 'Push unstable image', {
             def image = DockerImage.elifesciences(this, 'sciencebeam', commit)
             image.push()
```
Comment (de-code) [inline Jenkinsfile]: can we change the name to `sciencebeam_unstable` here?

Comment (de-code) [reply Jenkinsfile]: (also need to change downstream to pick up `sciencebeam_unstable` instead - but maybe it's better to run the texture demo based on the latest release instead)

File: Jenkinsfile [hunk 4]
```diff
@@ -10,6 +12,12 @@ elifeLibrary {
         stage 'Build images', {
             checkout scm
             dockerComposeBuild(commit)
+            candidateVersion = dockerComposeRunAndCaptureOutput(
```
Comment (de-code) [inline Jenkinsfile]: This is now also starting the dependency (GROBID) and its output is captured too. I guess the best is to remove the dependency via `docker-compose.ci.yml`

Comment (giorgiosironi) [reply Jenkinsfile]: I don't think you can remove entities through the additional file, only add options

Comment (giorgiosironi) [reply Jenkinsfile]: Maybe GROBID should write to stderr rather than stdout if it makes a difference?

File: Jenkinsfile (hunk 3 above)
Comment (giorgiosironi) [reply Jenkinsfile]: You should be able to change it here and the Docker Hub page should be auto-created

Comment (giorgiosironi) [reply Jenkinsfile]: However, the `DockerImage` class only works on existing objects so I think it needs some special method to also execute a `docker tag .../project:$commit .../project_unstable:$commit` command

File: Jenkinsfile (hunk 1 above)
Comment (giorgiosironi) [reply Jenkinsfile]: Good to separate, but I think unstable should go first as if there's something botched in the process it has a chance to fail on unstable first before touching the stable one.

File: Jenkinsfile (hunk 4 above)
Comment (de-code) [reply Jenkinsfile]: I guess I could also override the command

Comment (de-code) [reply Jenkinsfile]: Actually it's not the output from GROBID itself but docker.

Comment (de-code) [reply Jenkinsfile]: moved the `depends_on` to `docker-compose.override.yml` instead

File: Jenkinsfile (hunk 1 above)
Comment (de-code) [reply Jenkinsfile]: moved down

File: Jenkinsfile [hunk 5]
```diff
@@ -26,14 +34,23 @@ elifeLibrary {
...
+            isNew = sh(script: "git tag | grep v${candidateVersion}", returnStatus: true) != 0
+            if (isNew) {
+                def image = DockerImage.elifesciences(this, 'sciencebeam', commit)
+                image.tag(candidateVersion).push()
+            }
+        }
+
         stage 'Downstream', {
-            build job: '/dependencies/dependencies-sciencebeam-texture-update-sciencebeam', wait: false, parameters: [string(name: 'commit', value: commit)]
+            build job: '/dependencies/dependencies-sciencebeam-texture-update-sciencebeam', wait: false, parameters: [string(name: 'commit', value: candidateVersion)]
```
Comment (de-code) [inline Jenkinsfile]: passing down release version as `commit` which should work the same as it's really just the tag. Will rename parameter.

Comment (giorgiosironi) [reply Jenkinsfile]: Yes, I agree if there are releases available this should be wrapped with `if (isNew)`

**Comment by de-code** [comment]
I think it has everything in it now

File: Jenkinsfile (hunk 2 above)
Comment (giorgiosironi) [reply Jenkinsfile]: also these two could be swapped, first push `latest` and after that the release; if anything goes wrong you may have a `latest` without a corresponding release (allowed) but never a release that diverges from `latest`

Comment (giorgiosironi) [reply Jenkinsfile]: (same as "always push `master` before tags from `master`"

Comment (de-code) [reply Jenkinsfile]: We had that order before (push latest after commit). Having a latest without the release means something went wrong and maybe we shouldn't use it. But will comply.

Comment (giorgiosironi) [reply Jenkinsfile]: We should discuss some standard across projects, in general I tend to follow `commit` -> `master`|`latest` -> `tag`|release

Comment (giorgiosironi) [reply Jenkinsfile]: I agree in that case we shouldn't use it, because we would have pinned on the previous release tag

**Comment by giorgiosironi** [comment]
Will monitor what happens on the main pipeline.

**Pull request merged**